vernac/canonical.ml

```diff
@@ -11,29 +11,32 @@ open Names
 open Libobject
 open Recordops
 
-let open_canonical_structure i (_, o) =
+let open_canonical_structure i (_, (o,_)) =
   let env = Global.env () in
   let sigma = Evd.from_env env in
   if Int.equal i 1 then register_canonical_structure env sigma ~warn:false o
 
-let cache_canonical_structure (_, o) =
+let cache_canonical_structure (_, (o,_)) =
   let env = Global.env () in
   let sigma = Evd.from_env env in
   register_canonical_structure ~warn:true env sigma o
 
-let discharge_canonical_structure (_,x) = Some x
+let discharge_canonical_structure (_,(x, local)) =
+  match local with
+  | None | Some false -> Some (x, local)
+  | Some true -> None
 
-let inCanonStruc : Constant.t * inductive -> obj =
+let inCanonStruc : (Constant.t * inductive) * bool option -> obj =
```

Comment:
It would also work to not include the locality in the object, and instead do `if local && in_section then cache else add_leaf`` (ie skip the object when it's not useful)
Not sure if that's better.

Reply:
nope